File: code_eval/cpp/moderate/challenge_158.cpp
// Created by cschalk on 4/2/16.
// Challenge 158 : Interrupted Bubble Sort;
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>

using namespace std;

vector<int> parseSet(string setString) {
    vector<int> toReturn;
    stringstream ss(setString);
    int input;
    while (ss >> input) {
        toReturn.push_back(input);
    }
    return toReturn;
}

void runChallenge(string setString, long loops) {
    vector<int> set = parseSet(setString);
    long setSize = set.size();
    loops = min(setSize, loops);

    // Bubble Sort
    for (long i = 0 ; i < loops ; i++) {
        for (long j = 0 ; j < set.size() - 1; j++) {
            if (set[j] > set[j+1]) {
                int tmp = set[j];
                set[j] = set[j + 1];
                set[j + 1] = tmp;
            }
        }
    }

    for (int i : set) {
        cout << i << " ";
    }
    cout << endl;
}

int main (int argc, char ** argv) {
    ifstream input {argv[1]};
    string line;
    while(getline(input, line)) {
        int setTerminator = line.find_first_of('|');
        string setString = line.substr(0, setTerminator);
        line = line.substr(setTerminator + 1, line.length() - setTerminator);
        long loops = stol(line);

        runChallenge(setString, loops);
    }
    return 0;
}

File: code_eval/java/moderate/Challenge158.java
package moderate;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;

/**
 * Challenge 158 : Interrupted Bubble Sort
 * Created by cschalk on 3/21/16.
 */
public class Challenge158 {

    public static void main (String[] args) throws Exception {
        File file = new File(args[0]);
        BufferedReader buffer = new BufferedReader(new FileReader(file));
        String line;
        while ((line = buffer.readLine()) != null) {
             runChallenge(line);
        }
    }

    private static void runChallenge(String input) {
        String[] challenge = input.split("\\|");
        int[] toSort = parseChallengeString(challenge[0]);
        long loops = Long.parseLong(challenge[1].trim());
        int[] sorted = partiallyBubbleSort(toSort, loops);

        String out = "";
        for (Integer i : sorted) {
            out+=("" + i).concat(" ");
        }

        System.out.println(out.trim());
    }

    private static int[] parseChallengeString(String in){
        String[] list = in.split(" ");
        int [] toReturn = new int[list.length];
        for (int i = 0 ; i < list.length ; i++ ) {
            toReturn[i] = Integer.parseInt(list[i]);
        }
        return toReturn;
    }

    private static int[] partiallyBubbleSort(int[] list, long loops) {
        if (loops > list.length) loops = list.length;
        for (long i = 0 ; i < loops ; i++) {
            for (int j = 0 ; j < list.length - 1 ; j++) {
                if (list[j] > list[j+1]) {
                    int tmp = list[j];
                    list[j] = list [j + 1];
                    list[j + 1] = tmp;
                }
            }
        }
        return list;
    }
}
